src/ChangeStream.php

```diff
@@ -57,16 +57,19 @@ public function getId()
 
     public function key()
     {
-        return $this->key;
+        if ($this->valid() !== null) {
+            return $this->key;
+        }
+        return null;
     }
 
     public function next()
     {
-        $this->key++;
         try {
             $this->csIt->next();
             if ($this->valid()) {
                 $this->extractResumeToken($this->csIt->current());
+                $this->key++;
```

Comment:
I just noticed that ResumeTokenException extends the driver's RuntimeException class. If the user specifies a projection that removes the resume token, I think that would be caught by `catch (RuntimeException $e)` below and we'd resume unnecessarily. Resuming also wouldn't change the fact that a resume token is still being omitted from the result, so that might cause us to resume repeatedly.